userapi/src/main/java/model/D42Model.java

```diff
@@ -0,0 +1,117 @@
+package model;
+
+
+import fk.prof.storage.AsyncStorage;
+import fk.prof.storage.S3AsyncStorage;
+import org.apache.commons.codec.binary.Base32;
+import org.slf4j.LoggerFactory;
+
+import java.time.ZonedDateTime;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+/**
+ * Interacts with the AsyncStorage based on invocations from controller
+ * Created by rohit.patiyal on 19/01/17.
+ */
+public class D42Model implements IDataModel {
+    private static final String ACCESS_KEY = "66ZX9WC7ZRO6S5BSO8TG";
+    private static final String SECRET_KEY = "fGEJrdiSWNJlsZTiIiTPpUntDm0wCRV4tYbwu2M+";
+    private static final String S3_BACKUP_ELB_END_POINT = "http://10.47.2.3:80";
+    private static final String VERSION = "v001";
+    private static final String DELIMITER = "_";
+    private static final String BUCKET = "bck1";
+    private static final String PATH_SEPARATOR = "/";
+    private static final int TIMEOUT = 2;
+    private static org.slf4j.Logger LOGGER = LoggerFactory.getLogger(D42Model.class);
+    private S3AsyncStorage store = null;
+    private Base32 base32 = null;
+
+    public D42Model() {
+        base32 = new Base32();
+        setStorage(new S3AsyncStorage(S3_BACKUP_ELB_END_POINT, ACCESS_KEY, SECRET_KEY, Executors.newCachedThreadPool()));
+    }
+
+    private static String getWorkType(String key) {
+        String[] splits = key.split(DELIMITER);
+        return splits[6];
+    }
+
+    private static Profile getProfile(String key) {
+        String[] splits = key.split(DELIMITER);
+        String start = splits[4];
+        return new Profile(start, ZonedDateTime.parse(start).plusSeconds(Long.parseLong(splits[5])).toString());
+    }
+
+    private String getLastFromCommonPrefix(String commonPrefix) {
+        String[] splits = commonPrefix.split(DELIMITER);
+        return splits[splits.length - 1];
+    }
+
+    private Set<String> getListingAtLevelWithPrefix(String level, String objPrefix, boolean encoded) throws Exception {
+        Set<String> commonPrefixes = store.listAsync(level, false).get(TIMEOUT, TimeUnit.SECONDS);
```

Comment:
Change made by attaching a handler to the future, using `CompletableFuture.thenApply` which uses the same preceding stage's executor. It does the handling after the previous stage is compete.